Source/WebKit/NetworkProcess/storage/MemoryStorageArea.h

```diff
@@ -35,18 +35,22 @@ class StorageMap;
 
 namespace WebKit {
 
-class MemoryStorageArea final : public StorageAreaBase {
+class MemoryStorageArea final : public StorageAreaBase, public RefCounted<MemoryStorageArea> {
     WTF_MAKE_TZONE_ALLOCATED(MemoryStorageArea);
 public:
-    explicit MemoryStorageArea(const WebCore::ClientOrigin&, StorageAreaBase::StorageType = StorageAreaBase::StorageType::Session);
+    static Ref<MemoryStorageArea> create(const WebCore::ClientOrigin&, StorageAreaBase::StorageType = StorageAreaBase::StorageType::Session);
 
-    StorageAreaBase::Type type() const final { return StorageAreaBase::Type::Memory; };
+    StorageAreaBase::Type type() const final { return StorageAreaBase::Type::Memory; }
     bool isEmpty() final;
     void clear() final;
-    std::unique_ptr<MemoryStorageArea> clone() const;
+    Ref<MemoryStorageArea> clone() const;
     
+    void ref() const final { RefCounted::ref(); }
+    void deref() const final { RefCounted::deref(); }
```

Comment:
I believe this can be done with `using`, and doing so might even have a tiny efficiency effect.
```
using RefCounted::ref;
using RefCounted::deref;
```
I found that [cppreference.com](https://en.cppreference.com/w/cpp/language/using_declaration) says, "If the derived class already has a member with the same name, parameter list, and qualifications, the derived class member hides or overrides (doesn't conflict with) the member that is introduced from the base class."

I also noticed that we have macros named `DECLARE_VIRTUAL_REFCOUNTED` and `DEFINE_VIRTUAL_REFCOUNTED` that are apparently for this idiom. Maybe we should be using those in this pull request?

If so, then maybe later we can change them to the `using` technique.

Reply:
> I also noticed that we have macros named DECLARE_VIRTUAL_REFCOUNTED and DEFINE_VIRTUAL_REFCOUNTED that are apparently for this idiom. Maybe we should be using those in this pull request?

I am discussing those macros with Geoff (who introduced them) because I am not in favor of using them. I'd rather not hide ref() / deref() functions behind an obscure macro. This is also only 2 trivial lines so it doesn't feel worth the decreased readability / understandability from using the macro.

I'll look into `using` statements for this, I didn't realize this would work.

Reply:
Your proposal didn't build. I made this change:
```
diff --git a/Source/WebKit/NetworkProcess/storage/MemoryStorageArea.h b/Source/WebKit/NetworkProcess/storage/MemoryStorageArea.h
index 098a2d8d64b9..a2ecba929a65 100644
--- a/Source/WebKit/NetworkProcess/storage/MemoryStorageArea.h
+++ b/Source/WebKit/NetworkProcess/storage/MemoryStorageArea.h
@@ -45,8 +45,8 @@ public:
     void clear() final;
     Ref<MemoryStorageArea> clone() const;
     
-    void ref() const final { RefCounted::ref(); }
-    void deref() const final { RefCounted::deref(); }
+    using RefCounted::ref;
+    using RefCounted::deref;
 
 private:
     explicit MemoryStorageArea(const WebCore::ClientOrigin&, StorageAreaBase::StorageType);
```

and got:
```
/Volumes/Work/WebKit/OpenSource/Source/WebKit/NetworkProcess/storage/MemoryStorageArea.h:38:7: error: abstract class is marked 'final' [-Werror,-Wabstract-final-class]
   38 | class MemoryStorageArea final : public StorageAreaBase, public RefCounted<MemoryStorageArea> {
In file included from /Volumes/Work/WebKit/OpenSource/WebKitBuild/Debug/DerivedSources/WebKit/unified-sources/UnifiedSource36.cpp:1:
In file included from /Volumes/Work/WebKit/OpenSource/Source/WebKit/NetworkProcess/storage/SQLiteStorageArea.cpp:27:
In file included from /Volumes/Work/WebKit/OpenSource/Source/WebKit/NetworkProcess/storage/SQLiteStorageArea.h:28:
/Volumes/Work/WebKit/OpenSource/Source/WebKit/NetworkProcess/storage/StorageAreaBase.h:76:18: note: unimplemented pure virtual method 'ref' in 'MemoryStorageArea'
   76 |     virtual void ref() const = 0;
/Volumes/Work/WebKit/OpenSource/Source/WebKit/NetworkProcess/storage/StorageAreaBase.h:77:18: note: unimplemented pure virtual method 'deref' in 'MemoryStorageArea'
   77 |     virtual void deref() const = 0;
```

Reply:
Darn, I was wrong. Thanks for trying it.